Change default setting to `USE_BLAS64=0` when `USE_SYSTEM_BLAS==1`

Building Julia 0.3.6 from the latest release tarball with `USE_SYSTEM_BLAS=1` and `USE_SYSTEM_LAPACK=1` (among others) the test do not pass:

~~~~ 
   JULIA test/all
	From worker 3:	     * linalg2
	From worker 5:	     * linalg4
	From worker 2:	     * linalg1
	From worker 6:	     * linalg/cholmod
	From worker 7:	     * linalg/umfpack
	From worker 9:	     * keywordargs
	From worker 4:	     * linalg3
	From worker 8:	     * core
	From worker 6:	     * numbers
	From worker 9:	     * strings
exception on 4: ERROR: argument is not a permutation
 in invperm at ./combinatorics.jl:177
 in A_ldiv_B! at ./linalg/factorization.jl:413
 in A_ldiv_B! at ./linalg/factorization.jl:417
 in \ at ./linalg/factorization.jl:484
 in \ at ./linalg/dense.jl:417
 in anonymous at ./no file:9
 in runtests at /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading linalg3.jl, in expression starting on line 4
exception on 7: ERROR: assertion failed: |det(lua) - det(full(A))| <= 1.4210854715202004e-10
  det(lua) = 114.00000000000003
  det(full(A)) = -114.0
  difference = 228.00000000000003 > 1.4210854715202004e-10
 in error at ./error.jl:22
 in test_approx_eq at ./test.jl:109
 in anonymous at ./no file:22
 in runtests at /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading linalg/umfpack.jl, in expression starting on line 13
exception on 5: ERROR: ArgumentError("invalid argument #4294967296 to LAPACK call")
 in trtrs! at ./linalg/lapack.jl:1954
 in A_ldiv_B! at ./linalg/triangular.jl:47
 in \ at ./linalg/triangular.jl:49
 in \ at ./linalg/dense.jl:415
 in anonymous at ./no file:46
 in runtests at /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading linalg4.jl, in expression starting on line 29
exception on 2: ERROR: BoundsError()
 in ipiv2perm at ./linalg/lu.jl:87
 in getindex at ./linalg/lu.jl:99
 in cond at ./linalg/lu.jl:151
 in cond at ./linalg/dense.jl:451
 in anonymous at ./no file:12
 in runtests at /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading linalg1.jl, in expression starting on line 9
exception on 3: ERROR: assertion failed: |det(T) - det(F)| <= 0.0095367431640625
  det(T) = 8.618331
  det(F) = 0.0
  difference = 8.618331 > 0.0095367431640625
 in error at ./error.jl:22
 in test_approx_eq at ./test.jl:109
 in anonymous at ./no file:78
 in runtests at /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading linalg2.jl, in expression starting on line 22
ERROR: BoundsError()
 in anonymous at ./task.jl:1367
while loading linalg1.jl, in expression starting on line 9
while loading /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/runtests.jl, in expression starting on line 39

	From worker 3:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 8:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 5:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 4:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 2:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 6:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 9:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 7:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
~~~~

When I also pass `USE_ATLAS=1` as well as `ATLAS_LIBDIR=/path/to/atlas/lib` and `ATLAS_OBJ_SOURCE=/path/to/atlas/lib/libsatlas.so` I get different failures:

~~~
    JULIA test/all
	From worker 6:	     * linalg/cholmod
	From worker 7:	     * linalg/umfpack
	From worker 5:	     * linalg4
	From worker 2:	     * linalg1
	From worker 4:	     * linalg3
	From worker 3:	     * linalg2
	From worker 9:	     * keywordargs
	From worker 8:	     * core
	From worker 6:	     * numbers
	From worker 9:	     * strings
	From worker 7:	     * collections
	From worker 8:	     * hashing
	From worker 9:	     * remote
	From worker 9:	     * iobuffer
	From worker 9:	     * arrayops
	From worker 8:	     * reduce
	From worker 8:	     * reducedim
	From worker 7:	     * simdloop
	From worker 7:	     * blas
exception on 7: ERROR: test failed: all(BLAS.gemm('N','T',I4,I4) .== I4)
 in error at ./error.jl:21
 in default_handler at ./test.jl:19
 in do_test at ./test.jl:39
 in anonymous at ./no file:50
 in runtests at /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading blas.jl, in expression starting on line 3
exception on 2: ERROR: LAPACKException(5)
 in ggev! at ./linalg/lapack.jl:1380
 in eigfact! at ./linalg/factorization.jl:643
 in eigfact at ./linalg/factorization.jl:646
 in anonymous at ./no file:166
 in runtests at /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading linalg1.jl, in expression starting on line 25
ERROR: LAPACKException(5)
 in anonymous at ./task.jl:1367
while loading linalg1.jl, in expression starting on line 25
while loading /buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/test/runtests.jl, in expression starting on line 39

	From worker 8:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 5:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 7:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 2:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 4:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 3:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 6:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 9:	/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
/buildtmp/nix-build-julia-0.3.6.drv-4/julia-0.3/usr/bin/../lib/julia/sys.ji
Makefile:16: recipe for target 'all' failed
make[1]: *** [all] Error 1
Makefile:424: recipe for target 'test' failed
make: *** [test] Error 2
~~~~

I'm not sure if it is related, but in order to make the ccall mechanism work for the system libraries I build against I have patched `./src/ccall.cpp` to contain static entries in `sonameMap` for the following libraries:

~~~
libc
libpcre
libmpfr
libblas
liblapack
libgmp
libopenlibm
libopenspecfun
libfftw3
~~~

(See https://github.com/JuliaLang/julia/issues/10246 for a discussion on why this is necessary.)

Compiling with ATLAS and `USE_BLAS64=0` I get the same test failure (the second one in the post above).

Without ATLAS, i.e. just using lapack's libblas, and with `USE_BLAS64=0` I get this test failure instead:

~~~
    JULIA test/all
	From worker 2:	     * linalg1
	From worker 8:	     * core
	From worker 6:	     * linalg/cholmod
	From worker 3:	     * linalg2
	From worker 4:	     * linalg3
	From worker 5:	     * linalg4
	From worker 7:	     * linalg/umfpack
	From worker 9:	     * keywordargs
	From worker 6:	     * numbers
	From worker 9:	     * strings
	From worker 7:	     * collections
	From worker 8:	     * hashing
exception on 4: ERROR: ccall: could not find function cblas_cdotc_sub in library libblas
 in dotc at ./linalg/blas.jl:152
 in Ac_mul_B at ./linalg/matmul.jl:61
 in anonymous at ./no file:10
 in runtests at /tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/test/testdefs.jl:6
 in anonymous at ./multi.jl:855
 in run_work_thunk at ./multi.jl:621
 in anonymous at ./task.jl:855
while loading linalg3.jl, in expression starting on line 4
	From worker 9:	     * remote
ERROR: ccall: could not find function cblas_cdotc_sub in library libblas
 in anonymous at ./task.jl:1367
while loading linalg3.jl, in expression starting on line 4
while loading /tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/test/runtests.jl, in expression starting on line 39

	From worker 4:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 7:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 8:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 6:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 5:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 9:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 2:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
	From worker 3:	/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
/tmp/nix-build-julia-0.3.6.drv-0/julia-0.3/usr/bin/../lib/julia/sys.ji
Makefile:16: recipe for target 'all' failed
make[1]: *** [all] Error 1
Makefile:424: recipe for target 'test' failed
make: *** [test] Error 2
phase `check' failed after 115 seconds
~~~

The value of `Base.LinAlg.BlasInt` is in fact `Int64`.

--- a/Make.inc
+++ b/Make.inc
@@ -574,9 +574,17 @@ else
 PCRE_CONFIG = $(build_bindir)/pcre-config
 endif
 
-# Use 64-bit libraries by default on 64-bit architectures
+# Use ILP64 BLAS interface when building openblas from source on 64-bit architectures
 ifeq ($(BINARY), 64)
+ifeq ($(USE_SYSTEM_BLAS), 1)
+ifeq ($(USE_INTEL_MKL), 1)
 USE_BLAS64 ?= 1
+else # non MKL system blas is most likely LP64
+USE_BLAS64 ?= 0
+endif
+else
+USE_BLAS64 ?= 1
+endif
 endif
 
 ifeq ($(USE_SYSTEM_BLAS), 1)

--- a/contrib/windows/msys_build.sh
+++ b/contrib/windows/msys_build.sh
@@ -41,6 +41,7 @@ if [ "$ARCH" = x86_64 ]; then
   bits=64
   archsuffix=64
   echo "override MARCH = x86-64" >> Make.user
+  echo 'USE_BLAS64 = 1' >> Make.user
 else
   bits=32
   archsuffix=86

